src/games/oware.rs

```diff
@@ -0,0 +1,247 @@
+//!    pit is 2 or 3, then the current player captures it. It can happen several times if
+//!    continous multiple pits to the last dropped pit also have `2` or `3` seeds in them.
+//!    If all the seeds could be captured this turn, then no capture happens
+//!
+//! Objective is to capture as many seeds as possible and so the game will conclude early if a
+//! player captures more than half the seeds
+use std::fmt::{Debug, Display, Formatter};
+
+use internal_iterator::{Internal, IteratorExt};
+
+use crate::board::{Alternating, Board, BoardMoves, BruteforceMoveIterator, Outcome, Player};
+
+#[derive(Debug, Clone, Eq, PartialEq, Hash)]
+pub struct OwareBoard<const PITS: usize, const INIT_SEEDS: u8> {
+    pits: [u8; PITS],
+    scores: [u8; 2],
+    next_player: Player,
+    outcome: Option<Outcome>,
+}
+
+impl<const P: usize, const INIT_SEEDS: u8> Default for OwareBoard<P, INIT_SEEDS> {
+    fn default() -> Self {
+        OwareBoard {
+            pits: [INIT_SEEDS; P],
+            scores: Default::default(),
+            next_player: Player::A,
+            outcome: None,
+        }
+    }
+}
+
+impl<const PITS: usize, const I: u8> OwareBoard<PITS, I> {
+    fn is_in_opp_pits(&self, mv: usize) -> bool {
+        assert!(mv < PITS);
+        mv * 2 / PITS == usize::from(self.next_player == Player::A)
+    }
+    fn grand_slam(&self, mv: usize) -> bool {
+        self.is_in_opp_pits(mv)
+            && self.opp_pits().all(|x| match self.pits[x] {
+                2 | 3 => x <= mv,
+                0 => x > mv,
+                _ => false,
+            })
+            && self.pl_pits().any(|x| self.pits[x] > 0)
+    }
+    fn pl_pits(&self) -> std::ops::Range<usize> {
+        let pl = usize::from(self.next_player != Player::A) * PITS / 2;
+        pl..(pl + PITS / 2)
+    }
+    fn opp_pits(&self) -> std::ops::Range<usize> {
+        let pl = usize::from(self.next_player == Player::A) * PITS / 2;
+        pl..(pl + PITS / 2)
+    }
+    fn can_overflow(&self, mv: usize) -> bool {
+        mv % (PITS / 2) + self.pits[mv] as usize >= PITS / 2
+    }
+    fn is_final_move(&self) -> bool {
+        self.opp_pits().all(|x| self.pits[x] == 0) && !self.pl_pits().any(|x| self.can_overflow(x))
+    }
+    fn is_stalemate(&self) -> bool {
+        if self.pits.iter().sum::<u8>() == 2 && self.pits.iter().max() == Some(&1) {
+            let f = self.pits.iter().position(|&x| x == 1).unwrap();
+            let l = self.pits.iter().rposition(|&x| x == 1).unwrap();
+            l - f == PITS / 2
+        } else {
+            false
+        }
+    }
+}
+
+impl<const PITS: usize, const INIT_SEEDS: u8> Board for OwareBoard<PITS, INIT_SEEDS> {
+    type Move = usize;
+
+    fn next_player(&self) -> Player {
+        self.next_player
+    }
+
+    fn is_available_move(&self, mv: Self::Move) -> bool {
+        assert!(!self.is_done());
+        assert!(mv < PITS / 2);
+        assert!(PITS % 2 == 0, "const generic PITS should be even");
```

Comment:
I think it would be better to have the const generic parameter be `PITS_PER_PLAYER` instead (or something with a shorter name). That entirely removes this possible error condition, it's a bit awkward to handle it here at runtime.